### ISSUE TITLE
Project bounding box

### ISSUE THREAD

**Issue by navispatial** [navibyte/geospatial#96]
Add method like following to `Projection` mixin:

```dart
 R projectBox<R extends Box>(Box source, {required CreateBox<R> to});
```

Plan to calculate:
1. get distinct corners (one, two or four) in 2D for the source bbox
2. project all corner positions
3. calculate min and max coords in all axes from corner positions
4. handle special case of geographic coordinates when longitude west and east coordinates span anti-meridian
5. return the result as a new bounding box using a factory function

**Comment by navispatial** [comment]
Consider also bbox factory function:

```dart
/// Creates a new bounding box from [minX], [minY], [maxX] and [maxY] values.
///
/// Optional [minZ] and [maxZ] for 3D boxes, and [minM] and [maxM] for
/// measured boxes can be provided too.
///
/// For projected or cartesian bounding boxes (`ProjBox`), coordinates axis are
/// applied as is.
///
/// For geographic bounding boxes (`GeoBox`), coordinates are applied as:
/// `minX` => `west`, `minY` => `south`, `minZ` => `minElev`, `minM` => `minM`,
/// `maxX` => `east`, `maxY` => `north`, `maxZ` => `maxElev`, `maxM` => `maxM`
typedef CreateBox<T extends Box> = T Function({
  required num minX,
  required num minY,
  num? minZ,
  num? minM,
  required num maxX,
  required num maxY,
  num? maxZ,
  num? maxM,
});
```

**Comment by navispatial** [comment]
Implementatins on `ProjBox` and `GeoBox` would maybe be as follows.

`ProjBox`:

```dart
  /// A bounding box from parameters compatible with `CreateBox` function type.
  const ProjBox.create({
    required num minX,
    required num minY,
    num? minZ,
    num? minM,
    required num maxX,
    required num maxY,
    num? maxZ,
    num? maxM,
  })  : _minX = minX,
        _minY = minY,
        _minZ = minZ,
        _minM = minM,
        _maxX = maxX,
        _maxY = maxY,
        _maxZ = maxZ,
        _maxM = maxM;
```

`GeoBox`:

```dart
  /// A bounding box from parameters compatible with `CreateBox` function type.
  GeoBox.create({
    required num minX,
    required num minY,
    num? minZ,
    num? minM,
    required num maxX,
    required num maxY,
    num? maxZ,
    num? maxM,
  })  : _west = minX.toDouble(),
        _south = minY.toDouble(),
        _minElev = minZ?.toDouble(),
        _minM = minM?.toDouble(),
        _east = maxX.toDouble(),
        _north = maxY.toDouble(),
        _maxElev = maxZ?.toDouble(),
        _maxM = maxM?.toDouble();
```

**Comment by navispatial** [comment]
Implementation not quite as previously planned.

`GeoBox`:

```dart
  /// Projects this geographic bounding box to a projected box using
  /// the forward [projection].
  @override
  ProjBox project(Projection projection) {
    // get distinct corners (one, two or four) in 2D for the geographic bbox
    final corners = corners2D;

    // project all corner positions (using the forward projection)
    final projected = corners.map((pos) => pos.project(projection));

    // create a projected bbox
    // (calculating min and max coords in all axes from corner positions)
    return ProjBox.from(projected);
  }
```

`ProjBox`:

```dart
  /// Projects this projected bounding box to a geographic box using
  /// the inverse [projection].
  @override
  GeoBox project(Projection projection) {
    // get distinct corners (one, two or four) in 2D for the projected bbox
    final corners = corners2D;

    // unproject all corner positions (using the inverse projection)
    final unprojected = corners.map((pos) => pos.project(projection));

    // create an unprojected (geographic) bbox
    // (calculating min and max coords in all axes from corner positions)
    return GeoBox.from(unprojected);
  }
```

NOT handling *the special case of geographic coordinates when longitude west and east coordinates span anti-meridian*, this might need a separate issue.

**Comment by navispatial** [comment]
Implemented in [geobase 0.5.0](https://github.com/navibyte/geospatial/issues/174)
